File: src/events/messageReactionAdd.js
const { Events } = require('discord.js');
const fs = require('fs');
const path = require('path');
const { Suggestion } = require('../models/Suggestion');

module.exports = {
    name: Events.MessageReactionAdd,
    async execute(reaction, user) {
        // Ignorar reacciones de bots
        if (user.bot) return;

        // Asegurarse de que la reacción esté completamente cargada
        if (reaction.partial) {
            try {
                await reaction.fetch();
            } catch (error) {
                console.error('Error al cargar la reacción:', error);
                return;
            }
        }

        // Manejar roles por reacción (múltiples mappings)
        if (reaction.client.reactionRoles && reaction.client.reactionRoles.has(reaction.message.id)) {
            const cfg = reaction.client.reactionRoles.get(reaction.message.id);
            const mappings = cfg.mappings || [];
            for (const mapping of mappings) {
                try {
                    if (reaction.emoji.toString() === mapping.emoji) {
                        const guild = reaction.message.guild;
                        const member = await guild.members.fetch(user.id);
                        const role = guild.roles.cache.get(mapping.roleId);
                        if (role) await member.roles.add(role);
                    }
                } catch (error) {
                    console.error('Error al procesar mapping de reactionrole:', error);
                }
            }
        }

        // Manejar reacciones en sugerencias para actualizar contador de aprobaciones
        try {
            const sugg = await Suggestion.findOne({ messageId: reaction.message.id }).exec();
            if (sugg) {
                // Si la reacción añadida es la de aprobación y quien reaccionó es el autor de la sugerencia,
                // remover la reacción y no contar ese voto.
                const isUpvote = (reaction.emoji && (reaction.emoji.name === '👍' || reaction.emoji.toString() === '👍'));
                if (isUpvote && sugg.authorId === user.id) {
                    try {
                        // Intentar remover la reacción del autor. Requiere permiso para MANAGE_MESSAGES en el canal.
                        await reaction.users.remove(user.id).catch(() => {});
                    } catch (remErr) {
                        // No interrumpir el flujo si no se pudo remover la reacción.
                        console.error('No se pudo remover la reacción del autor de la sugerencia:', remErr.message || remErr);
                    }
                    // No contamos ni actualizamos nada puesto que la reacción del autor fue ignorada/quitada.
                    return;
                }

                const upReaction = reaction.message.reactions.cache.get('👍');
                let count = 0;
                if (upReaction) {
                    const users = await upReaction.users.fetch();
                    count = users.filter(u => !u.bot && u.id !== sugg.authorId).size;
                }
                sugg.approvals = count;
                await sugg.save().catch(e => console.error('No se pudo guardar sugerencia en MongoDB', e));

                // actualizar embed (preservar color/imagen basados en el estado guardado)
                const embed = reaction.message.embeds[0] ? reaction.message.embeds[0].toJSON() : null;
                if (embed) {
                    const { EmbedBuilder } = require('discord.js');
                    const e = EmbedBuilder.from(embed);

                    // Actualizar votos
                    const upvotes = reaction.message.reactions.cache.get('👍')?.count || 0;
                    const downvotes = reaction.message.reactions.cache.get('👎')?.count || 0;

                    // Actualizar campos preservando el formato
                    e.data.fields = embed.fields.map(f => {
                        if (f.name === 'Votos') {
                            return {
                                name: 'Votos',
                                value: `👍 ${upvotes - 1} | 👎 ${downvotes - 1}`,
                                inline: true
                            };
                        }
                        return f;
                    });

                    // Determinar color por estado
                    try {
                        if (sugg.status === 'Aprobada' || sugg.status === 'Implementada') e.setColor('#2ECC71');
                        else if (sugg.status === 'Denegada') e.setColor('#E74C3C');
                        else e.setColor('#3498db');
                    } catch (ce) { /* ignore */ }

                    // Asegurar thumbnail del autor
                    if (sugg.authorAvatar) {
                        try { e.setThumbnail(sugg.authorAvatar); } catch (ie) { /* ignore */ }
                    }

                    await reaction.message.edit({ embeds: [e] }).catch(() => null);
                }
            }
        } catch (err) {
            console.error('Error actualizando aprobaciones de sugerencia:', err);
        }
    },
};

File: pybot/events/message_reaction_add.py
from __future__ import annotations

import discord
from discord.ext import commands

from ..database import is_mongo_connected, suggestions_repository


class MessageReactionAddEvents(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @commands.Cog.listener()
    async def on_reaction_add(self, reaction: discord.Reaction, user: discord.User) -> None:
        if user.bot:
            return
        if reaction.message.partial:
            try:
                await reaction.message.fetch()
            except Exception:
                return

        if not await is_mongo_connected():
            return

        record = await suggestions_repository.get_suggestion_by_message(reaction.message.id)
        if not record:
            return

        is_upvote = str(reaction.emoji) == "👍" or getattr(reaction.emoji, "name", None) == "👍"
        if is_upvote and record.author_id == str(user.id):
            try:
                await reaction.remove(user)
            except Exception:
                pass
            return

        await update_suggestion_reactions(reaction.message, record)


async def update_suggestion_reactions(message: discord.Message, record) -> None:
    up_reaction = discord.utils.get(message.reactions, emoji="👍")
    down_reaction = discord.utils.get(message.reactions, emoji="👎")
    upvotes = 0
    downvotes = 0
    if up_reaction:
        try:
            users = await up_reaction.users().flatten()
            upvotes = len([u for u in users if not u.bot and str(u.id) != str(record.author_id)])
        except Exception:
            upvotes = max(0, (up_reaction.count or 0) - 1)
    if down_reaction:
        try:
            users = await down_reaction.users().flatten()
            downvotes = len([u for u in users if not u.bot])
        except Exception:
            downvotes = max(0, (down_reaction.count or 0) - 1)

    await suggestions_repository.update_approvals(record.id, upvotes)

    if not message.embeds:
        return
    embed_dict = message.embeds[0].to_dict()
    fields = []
    for field in embed_dict.get("fields", []):
        name = field.get("name")
        if name == "Votos":
            field["value"] = f"👍 {upvotes} | 👎 {downvotes}"
        fields.append(field)
    embed_dict["fields"] = fields
    colour = config_colour_for_status(record.status)
    embed_dict["color"] = colour.value
    updated = discord.Embed.from_dict(embed_dict)
    if record.author_avatar:
        updated.set_thumbnail(url=record.author_avatar)
    try:
        await message.edit(embed=updated)
    except Exception:
        pass


def config_colour_for_status(status: str) -> discord.Colour:
    status = (status or "").lower()
    if status in {"aprobada", "implementada"}:
        return discord.Colour(0x2ECC71)
    if status == "denegada":
        return discord.Colour(0xE74C3C)
    return discord.Colour(0x3498DB)


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(MessageReactionAddEvents(bot))
